pkg/cloudprovider/metal/instances_v2.go

```diff
@@ -143,3 +147,53 @@ func (o *metalInstancesV2) InstanceMetadata(ctx context.Context, node *corev1.No
 		Region:        region,
 	}, nil
 }
+
+func (o *metalInstancesV2) getServerClaimForNode(ctx context.Context, node *corev1.Node) (*metalv1alpha1.ServerClaim, error) {
+	var serverClaim *metalv1alpha1.ServerClaim
+	if node.Spec.ProviderID != "" {
+		return o.getServerClaimFromProviderID(ctx, node.Spec.ProviderID)
+	}
+
+	serverClaimList := &metalv1alpha1.ServerClaimList{}
+	if err := o.metalClient.List(ctx, serverClaimList, client.InNamespace(o.metalNamespace)); err != nil {
+		return nil, fmt.Errorf("failed to list server claims for node %s: %w", node.Name, err)
+	}
+
+	for _, claim := range serverClaimList.Items {
+		server := &metalv1alpha1.Server{}
+		if err := o.metalClient.Get(ctx, client.ObjectKey{Name: claim.Spec.ServerRef.Name}, server); err != nil {
+			return nil, fmt.Errorf("failed to get server object for node %s: %w", node.Name, err)
+		}
+		if nodeInfo := node.Status.NodeInfo; nodeInfo.SystemUUID == server.Spec.UUID {
+			return &claim, nil
+		}
+	}
+
+	return serverClaim, nil
+}
+
+func (o *metalInstancesV2) getServerClaimFromProviderID(ctx context.Context, providerID string) (*metalv1alpha1.ServerClaim, error) {
+	objKey, err := getObjectKeyFromProviderID(providerID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get object key for ProviderID %s: %w", providerID, err)
+	}
+	serverClaim := &metalv1alpha1.ServerClaim{}
+	if err := o.metalClient.Get(ctx, objKey, serverClaim); err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, cloudprovider.InstanceNotFound
+		}
+		return nil, fmt.Errorf("failed to get server claim object for ProviderID %s: %w", providerID, err)
+	}
+	return serverClaim, nil
+}
+
+func getObjectKeyFromProviderID(providerID string) (client.ObjectKey, error) {
+	parts := strings.Split(strings.TrimPrefix(providerID, fmt.Sprintf("%s://", ProviderName)), "/")
+	if len(parts) != 2 {
```

Comment:
Should we also keep ServerUUID at the end, of the ProviderID? Then ProviderID can uniquely identify the machine even across regions etc.